Automatically use ReleaseExe when top-level statements are present in a compilation and no options are specified.

333fred opened the pull request.

333fred commented:
@dotnet/roslyn-compiler for a quick review of this test-only PR that we discussed in the meeting today.

333fred commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -799,7 +799,7 @@ public void M1()
         public void ImplicitObjectCreationUnconverted_ArrayIndex()
         {
-            var comp = CreateCompilation("_ = new int[/*<bind>*/new(bad)/*</bind>*/];", options: TestOptions.ReleaseExe);
+            var comp = CreateCompilation("_ = new int[/*<bind>*/new(bad)/*</bind>*/];");
```
Removed a couple of these from tests I'd recently created to make sure it was working.

jcouv reviewed (approved):
LGTM Thanks (iteration 1)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -82,7 +82,7 @@ class Point(int x, int y);
                 Diagnostic(ErrorCode.ERR_UseDefViolation, "int y").WithArguments("y").WithLocation(2, 20)
             );
-            comp = CreateCompilation(src2, parseOptions: TestOptions.Regular8);
+            comp = CreateCompilation(src2, parseOptions: TestOptions.Regular8, options: TestOptions.ReleaseDll);
```
>, options: TestOptions.ReleaseDll [](start = 77, length = 33)

Is this change necessary?

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs:
```diff
@@ -293,7 +293,7 @@ public class B
             var countedTree = new CountedSyntaxTree(foreignType);
 
-            var compilation = CreateCompilation(new SyntaxTree[] { underlyingTree, countedTree }, skipUsesIsNullable: true);
+            var compilation = CreateCompilation(new SyntaxTree[] { underlyingTree, countedTree }, skipUsesIsNullable: true, options: TestOptions.ReleaseDll);
```
>, options: TestOptions.ReleaseDll [](start = 122, length = 33)

Is this change necessary?

333fred replied:
Yes, it changes all the `AccessCounts` to 2 instead of 1.

---
In reply to: [554043500](https://github.com/dotnet/roslyn/pull/50325#discussion_r554043500) [](ancestors = 554043500)

AlekseyTs replied:
It feels suspicious that changing the output kind would have this effect. I think we need to understand what is going on.

---
In reply to: [554099386](https://github.com/dotnet/roslyn/pull/50325#discussion_r554099386) [](ancestors = 554099386,554043500)

333fred replied:
Ah, ok. It's because calls to `GetRoot()` are counted, so the new check for "does this syntax have top-level statements" increments the counter. I'll add a remark to `CountedSyntaxTree` about it.

---
In reply to: [554109951](https://github.com/dotnet/roslyn/pull/50325#discussion_r554109951) [](ancestors = 554109951,554099386,554043500)

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:
```diff
@@ -1090,9 +1091,13 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
             if (options == null)
             {
-                options = TestOptions.ReleaseDll;
+                bool hasTopLevelStatements = syntaxTrees.Any(s => s.GetRoot().DescendantNodes().OfType<GlobalStatementSyntax>().Any());
```
>DescendantNodes() [](start = 78, length = 17)

Does this look too deep? I think we should check only immediate children of a compilation unit.
See
```
        internal static bool IsTopLevelStatement([NotNullWhen(true)] GlobalStatementSyntax? syntax)
        {
            return syntax?.Parent?.IsKind(SyntaxKind.CompilationUnit) == true;
        }
```

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [756852938](https://github.com/dotnet/roslyn/pull/50325#issuecomment-756852938) [](ancestors = 756852938)

AlekseyTs reviewed (approved):
LGTM (commit 2)
